packages/react-reconciler/src/ReactFiberWorkLoop.new.js

```diff
@@ -1250,16 +1260,21 @@ export function flushSync<A, R>(fn: A => R, a: A): R {
     return fn(a);
   }
   executionContext |= BatchedContext;
-  const previousLanePriority = getCurrentUpdateLanePriority();
+  let previousLanePriority;
   try {
-    setCurrentUpdateLanePriority(SyncLanePriority);
+    if (enableSetUpdateLanePriority) {
+      previousLanePriority = getCurrentUpdateLanePriority();
+      setCurrentUpdateLanePriority(SyncLanePriority);
+    }
     if (fn) {
       return runWithPriority(ImmediateSchedulerPriority, fn.bind(null, a));
     } else {
       return (undefined: $FlowFixMe);
     }
   } finally {
-    setCurrentUpdateLanePriority(previousLanePriority);
+    if (enableSetUpdateLanePriority && previousLanePriority != null) {
```

Comment:
Factoring nit: Move the flag check outside the try-catch. Then you don't have to do the weird `previousLanePriority` check.

Something like:

```js
if (fn) {
  if (enableSetUpdateLanePriority) {
    return runWithPriority(ImmediateSchedulerPriority, fn.bind(null, a));
  } else {
    const previousLanePriority = getCurrentUpdateLanePriority();
    setCurrentUpdateLanePriority(SyncLanePriority);
    try {
      return runWithPriority(ImmediateSchedulerPriority, fn.bind(null, a));
    } finally {
      setCurrentUpdateLanePriority(previousLanePriority);
    }
  }
}
```

I find this easier to see the difference in behavior.

Reply:
+1